godot-cell/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+    _pin: PhantomPinned,
+}
+
+impl<T> GdCell<T> {
+    /// Creates a new cell storing `value`.
+    pub fn new(value: T) -> Pin<Box<Self>> {
+        let cell = Box::pin(Self {
+            state: Mutex::new(CellState::new()),
+            value: UnsafeCell::new(value),
+            _pin: PhantomPinned,
+        });
+
+        cell.state.lock().unwrap().initialize_ptr(&cell.value);
+
+        cell
+    }
+
+    /// Returns a new shared reference to the contents of the cell.
+    ///
+    /// Fails if an aliasing mutable reference exists.
+    pub fn borrow(self: Pin<&Self>) -> Result<RefGuard<'_, T>, Box<dyn Error>> {
+        let mut state = self.state.lock().unwrap();
+        state.borrow_state.increment_shared()?;
+
+        // SAFETY:
+        // `increment_shared` succeeded, therefore there cannot currently be any aliasing mutable references.
+        unsafe { Ok(RefGuard::new(&self.get_ref().state, state.get_ptr())) }
+    }
+
+    /// Returns a new shared reference to the contents of the cell.
+    ///
+    /// Fails if an aliasing mutable reference exists, or a shared reference exists.
+    pub fn borrow_mut(self: Pin<&Self>) -> Result<MutGuard<'_, T>, Box<dyn Error>> {
+        let mut state = self.state.lock().unwrap();
+        state.borrow_state.increment_mut()?;
+        let count = state.borrow_state.mut_count();
+        let value = state.get_ptr();
+
+        // SAFETY:
+        // `increment_mut` succeeded, therefore any existing mutable references do not alias, and no new
+        // references may be made unless this one is guaranteed not to alias those.
+        //
+        // This is the case because the only way for a new `GdMut` or `GdRef` to be made after this, is for
+        // either this guard to be dropped or `set_non_aliasing` to be called.
+        //
+        // If this guard is dropped, then we dont need to worry.
+        //
+        // If `set_non_aliasing` is called, then either a mutable reference from this guard is passed in.
+        // In which case, we cannot use this guard again until the resulting non-aliasing guard is dropped.
+        //
+        // We cannot pass in a different mutable reference, since `set_non_aliasing` ensures any references
+        // matches the ones this one would return. And only one mutable reference to the same value can exist
+        // since we cannot have any other aliasing mutable references around to pass in.
+        unsafe { Ok(MutGuard::new(&self.get_ref().state, count, value)) }
+    }
+
+    /// Set the current mutable borrow as not aliasing any other references.
+    ///
+    /// Will error if there is no current possibly aliasing mutable borrow, or if there are any shared
+    /// references.
+    pub fn set_non_aliasing<'a, 'b>(
+        self: Pin<&'a Self>,
+        current_ref: &'b mut T,
+    ) -> Result<NonAliasingGuard<'b, T>, Box<dyn Error>>
+    where
+        'a: 'b,
+    {
```

Comment:
Do you think it would make sense to name the lifetimes?

E.g. `'cell` and `'val` or so...